feat(conductor)!: fetch blocks pending finalization

SuperFluffy opened the pull request:
## Summary

Adds the ability for Conductor to sync its connected
rollup if firm commitments trail soft commitments.

## Background
Conductor uses the
`astria.execution.v1alpha2.BatchGetBlocks` API to
retrieve all blocks in the range `firm+1..=soft`
from the rollup node. For example, if the firm
number is 3 and soft number is 10, then all
soft-committed blocks 4 through 10 (inclusive) are
fetched.

Previously conductor would fail try to re-execute
the firm blocks as it didn't have their
corresponding rollup-hashes, causing the rollup
node to reject the execution.

## Changes

1. Conductor now matches rollup block numbers
rollup blocks (previously, rollup block hashes,
which it doesn't have on a fresh start).
2. Conductor uses these block numbers to update
the commitment state upon receiving a pending
firm block.
3. Reworked the channel tracking the rollup state
in the executor task: since Conductor must now be
able to in both directions between Sequencer
heights and Rollup numbers, updating the tracked
commitment state is now a fallible operation. This
required an abstraction on top of the
watch-channel containing the state.
4. changed how the next expected Sequencer height
is calculated from the Rollup number (for both
firm and soft commitments). It is now
sequencer_genesis_block_height + block_number + 1
(previously without + 1, which was confusing).

## Testing

1. Added 2 blackbox tests for soft-and-only mode,
one for testing the pending-block fetch.
2. Removed the executor tests as they are now covered
by the blackbox tests.

## Breaking Changelist

1. The calculation of the next expected Sequencer
height is now one more than previous. This is a
breaking change in the sense that rollup + conductor
need to started with the same view of
rollup number <> sequencer height in mind. A rollup
that was started with an older conductor will not work
with this conductor.

## Related Issues
Link any issues that are related, prefer full github links.

closes https://github.com/astriaorg/astria/issues/837

joroshiba commented on crates/astria-conductor/src/executor/mod.rs:
```diff
@@ -437,8 +513,60 @@ impl Executor {
+        &mut self,
+        mut client: Client,
+    ) -> eyre::Result<()> {
+        if self.firm_blocks.is_none() || self.soft_blocks.is_none() {
```
should this be an error case? Why would we call this before setting soft and firm block?

SuperFluffy replied:
`firm_blocks` and `soft_blocks` are the channels that are created during executor construction.

We use these as proxies for which "mode" executor operates under.

joroshiba commented on crates/astria-conductor/src/executor/mod.rs:
```diff
@@ -437,8 +513,60 @@ impl Executor {
+            .await
+            .wrap_err("failed getting blocks for not yet finalized firm heights")?;
+
+        info!(
+            pending_blocks = %json(&reporting::ReportBlocks(&blocks)),
+            "received blocks pending finalization",
+        );
```
this could be a lot of blocks? We probably don't want to have an info log with all of them? IE if variance window is 100, 600 blocks?

SuperFluffy replied:
You are right, this is not actionable information. I have removed the `pending_blocks` field entirely. The `info!` event now leans on the `start` and `end` block numbers that are part of the `populate_blocks_pending_finalization` span.

joroshiba commented on crates/astria-conductor/src/executor/reporting.rs:
```diff
@@ -0,0 +1,20 @@
+    where
+        S: serde::Serializer,
+    {
+        let mut seq = serializer.serialize_seq(Some(self.0.len()))?;
```
again concerned about length here

SuperFluffy replied:
Removed as no longer used.

joroshiba reviewed (changes requested):
only minors, but concerned about the logging

noot commented on crates/astria-conductor/tests/blackbox/soft_and_firm.rs:
```diff
@@ -0,0 +1,318 @@
+         within 1000ms",
+    );
+
+    // mount_get_filtered_sequencer_block!(
```
remove

noot commented on crates/astria-conductor/src/executor/mod.rs:
```diff
@@ -108,84 +142,94 @@ impl Handle<StateIsInit> {
     /// Required to mark firm blocks received from celestia as executed
     /// without re-executing on top of the rollup node on top of the rollup node..
-    blocks_pending_finalization: HashMap<[u8; 32], Block>,
+    blocks_pending_finalization: HashMap<u32, Block>,
```
is this rollup_block_height to block?

joroshiba replied:
Would be good to have a comment above on what the height is for next person through.

SuperFluffy replied:
Yes, this was confusing. Changed to:
```
    /// Tracks rollup blocks by their rollup block numbers.
    ///
    /// Required to mark firm blocks received from celestia as executed
    /// without re-executing on top of the rollup node.
```

noot commented on crates/astria-conductor/src/executor/tests.rs:
```diff
@@ -1,554 +1,27 @@
-}
-
-#[tokio::test]
-async fn firm_blocks_at_expected_heights_are_executed() {
```
why are all these deleted?

noot commented on crates/astria-conductor/tests/blackbox/soft_and_firm.rs:
```diff
@@ -0,0 +1,318 @@
+    )
+    .await
+    .expect(
+        "conductor should have executed the soft block and updated the soft commitment state \
```
```suggestion
        "conductor should have executed the block and updated the soft and firm commitment state \
```

SuperFluffy commented on crates/astria-conductor/src/executor/mod.rs:
```diff
@@ -272,25 +328,23 @@ impl Executor {
     /// that Conductor will run on.
-    fn calculate_max_spread(&self) -> usize {
-        usize::try_from(self.state.borrow().celestia_block_variance())
+    fn calculate_max_spread(&mut self) -> usize {
```
This actually begs the question: why batch-request all these blocks in the first place? We could just as well request the parent-block from the rollup as we go instead of maintaing a cache.

SuperFluffy commented on crates/astria-conductor/src/executor/mod.rs:
```diff
@@ -369,52 +430,67 @@ impl Executor {
+                    "did not find pending block; executing firm block"
+                );
+                // The parent hash of the next block is the hash of the block at the current head.
+                let parent_hash = self.state.firm_hash();
```
The entire logic is faulty: if `sort > firm` and there is no soft block number, then that's a hard error and the block must not be executed.

As you note, this code is not new but can be traced back to the original implementation last year. I am going to address this in a follow-up PR.

(Actually, this begs the question: why batch-request all these blocks in the first place? We could just as well request the parent-block from the rollup as we go instead of maintaining a cache. I am going to leave the implementation as it is for now, but just food for thought.)

SuperFluffy replied:
https://github.com/astriaorg/astria/issues/1011

SuperFluffy commented:
> only minors, but concerned about the logging

Addressed all comments. In particular I added the following new code:

1. `executor::Client::batch_get_blocks` now ensures that the rollup response does not violate the "batch" contract (no holes, no dupes, no extra blocks outside the requested range)
2. the event no longer emits all blocks that the batch-get-block RPC returned. It only emits a message now (start and end numbers are already part of the span that the event is under)

noot reviewed (approved):
looks good!

joroshiba reviewed (approved):
I don't think fixing the contract code is necessarily blocking, but it can be easily simplified and I don't think validates the actual contract. In this instance we are never requesting dupes, but we aren't validating order matches. Makes it less useful from a "make sure my rollup is behaiving correctly" sense.

SuperFluffy commented on crates/astria-conductor/src/executor/client.rs:
```diff
@@ -168,3 +183,119 @@ impl Client {
+    Extras(HashSet<u32>),
+}
+
+fn ensure_batch_get_blocks_is_correct(
```
This was a good point. I have changed simplified the code accordingly and amended the tests.

SuperFluffy commented on crates/astria-conductor/src/executor/client.rs:
```diff
@@ -168,3 +183,119 @@ impl Client {
+    mut blocks: Vec<RawBlock>,
+    requested_numbers: RangeInclusive<u32>,
+) -> Result<Vec<RawBlock>, BatchGetBlocksError> {
+    blocks.sort_unstable_by_key(|block| block.number);
```
Contract is now actually checked! :)
